submit/controllers/cross.py

```diff
@@ -0,0 +1,214 @@
+
+from flask import url_for, Markup
+from wtforms import Form, widgets
+from wtforms.fields import Field, BooleanField, HiddenField
+from wtforms.validators import InputRequired, ValidationError, optional, \
+    DataRequired
+
+from arxiv import status
+from arxiv.taxonomy import CATEGORIES_ACTIVE as CATEGORIES
+from arxiv.taxonomy import ARCHIVES_ACTIVE as ARCHIVES
+from arxiv.base import logging, alerts
+from arxiv.forms import csrf
+from arxiv.users.domain import Session
+import arxiv.submission as events
+
+from ..util import load_submission
+from . import util
+
+logger = logging.getLogger(__name__)  # pylint: disable=C0103
+
+Response = Tuple[Dict[str, Any], int, Dict[str, Any]]  # pylint: disable=C0103
+
+
+CONTACT_SUPPORT = Markup(
+    'If you continue to experience problems, please contact'
+    ' <a href="mailto:help@arxiv.org"> arXiv support</a>.'
+)
+
+
+class HiddenListField(HiddenField):
+    def process_formdata(self, valuelist):
+        self.data = list(str(x) for x in valuelist if x)
+
+    def process_data(self, value):
+        try:
+            self.data = list(str(v) for v in value if v)
+        except (ValueError, TypeError):
+            self.data = None
+
+    def _value(self):
+        return ",".join(self.data) if self.data else ""
+
+
+class CrossListForm(csrf.CSRFForm):
+    """Submit a cross-list request."""
+
+    CATEGORIES = [
+        (archive['name'], [
+            (category_id, f"{category['name']} ({category_id})")
+            for category_id, category in CATEGORIES.items()
+            if category['in_archive'] == archive_id
+        ])
+        for archive_id, archive in ARCHIVES.items()
+    ]
+    """Categories grouped by archive."""
+
+    ADD = 'add'
+    REMOVE = 'remove'
+    OPERATIONS = [
+        (ADD, 'Add'),
+        (REMOVE, 'Remove')
+    ]
+    operation = HiddenField(default=ADD, validators=[optional()])
+    category = util.OptGroupSelectField('Category', choices=CATEGORIES,
+                                        default='', validators=[optional()])
+    selected = HiddenListField()
+    confirmed = BooleanField('Confirmed',
+                             false_values=('false', False, 0, '0', ''))
+
+    def validate_selected(form: csrf.CSRFForm, field: Field) -> None:
+        if form.confirmed.data and not field.data:
+            raise ValidationError('Please select a category')
+        for value in field.data:
+            if value not in CATEGORIES:
+                raise ValidationError('Not a valid category')
+
+    def validate_category(form: csrf.CSRFForm, field: Field) -> None:
+        if not form.confirmed.data and not field.data:
+            raise ValidationError('Please select a category')
+
+    def filter_choices(self, submission: events.domain.Submission,
+                       session: Session,
+                       exclude: Optional[List[str]] = None) -> None:
+        """Remove redundant choices, and limit to endorsed categories."""
```

Comment:
After my discussion with Martin about endorsements, as currently implemented there might not be such a thing as 'limit to endorsed categories'? This would be because a user who has the ability to be auto-endorsed but has never selected that particular category doesn't have an entry in the classic DB for the selected category, it gets generated during the checking of endorsement status. 
This list would definitely be able to be 'limit to user-selected categories' though.

Am I missing something about how endorsements gets implemented in NG?

Reply:
We've actually worked around this, and are inferring the auto-endorsed categories before they are actually auto-endorsed. This is implemented here: https://github.com/cul-it/arxiv-auth/blob/develop/users/arxiv/users/legacy/endorsements.py . The short story is that the endorsements on the user on the session should represent all of the possible explicitly and would-be-auto-endorsed categories.